assets/js/amp-wp-app-shell.js

```diff
@@ -0,0 +1,409 @@
+			return;
+		}
+
+		// Skip handling click if it was handled already.
+		if ( event.defaultPrevented ) {
+			return;
+		}
+
+		const url = new URL( event.target.action );
+		if ( ! isLoadableURL( url ) ) {
+			return;
+		}
+
+		for ( const element of event.target.elements ) {
+			if ( element.name && ! element.disabled ) {
+				// @todo Need to handle radios, checkboxes, submit buttons, etc.
+				url.searchParams.set( element.name, element.value );
+			}
+		}
+		loadUrl( url, { scrollIntoView: true } );
+		event.preventDefault();
+	}
+
+	/**
+	 * Determine whether header is visible at all.
+	 *
+	 * @return {boolean} Whether header image is visible.
+	 */
+	function isHeaderVisible() {
+		const element = document.querySelector( '.site-branding' );
+		const clientRect = element.getBoundingClientRect();
+		return clientRect.height + clientRect.top >= 0;
+	}
+
+	/**
+	 * Fetch response for URL of shadow doc.
+	 *
+	 * @param {string} url - URL.
+	 * @return {Promise<Response>} Response promise.
+	 */
+	function fetchShadowDocResponse( url ) {
+		const ampUrl = new URL( url );
+		ampUrl.searchParams.set( ampAppShell.componentQueryVar, 'inner' );
+
+		return fetch( ampUrl.toString(), {
+			method: 'GET',
+			mode: 'same-origin',
+			credentials: 'include',
+			redirect: 'follow',
+			cache: 'default',
+			referrer: 'client'
+		} );
+	}
+
+	/**
+	 * Load URL.
+	 *
+	 * @todo When should scroll to the top? Only if the first element of the content is not visible?
+	 * @param {string|URL} url - URL.
+	 * @param {boolean} scrollIntoView - Scroll into view.
+	 * @param {boolean} pushState - Whether to push state.
+	 */
+	function loadUrl( url, { scrollIntoView = false, pushState = true } = {} ) {
+		const previousUrl = location.href;
+		const navigateEvent = new CustomEvent( 'wp-amp-app-shell-navigate', {
+			cancelable: true,
+			detail: {
+				previousUrl,
+				url: String( url )
+			}
+		} );
+		if ( ! window.dispatchEvent( navigateEvent ) ) {
+			return; // Allow scripts on page to cancel navigation via preventDefault() on wp-amp-app-shell-navigate event.
+		}
+
+		updateNavMenuClasses( url );
+
+		fetchShadowDocResponse( url )
+			.then( response => {
+				if ( currentShadowDoc ) {
+					currentShadowDoc.close();
+				}
+
+				const oldContainer = document.getElementById( ampAppShell.contentElementId );
+				const newContainer = document.createElement( oldContainer.nodeName );
+				newContainer.setAttribute( 'id', oldContainer.getAttribute( 'id' ) );
+				oldContainer.parentNode.replaceChild( newContainer, oldContainer );
+
+				/*
+				 * For more on this, see:
+				 * https://www.ampproject.org/latest/blog/streaming-in-the-shadow-reader/
+				 * https://github.com/ampproject/amphtml/blob/master/spec/amp-shadow-doc.md#fetching-and-attaching-shadow-docs
+				 */
+				currentShadowDoc = AMP.attachShadowDocAsStream( newContainer, url, {} );
```

Comment:
@westonruter The `url` parameter here comes through as an [URL object](https://developer.mozilla.org/en-US/docs/Web/API/URL) when the navigation happens through a click or form submit, but the method [expects](https://github.com/ampproject/amphtml/blob/f41de5e5205ae37a823e774af817e2748ad5ed09/src/multidoc-manager.js#L273) and [newly enforces](https://github.com/ampproject/amphtml/commit/f41de5e5205ae37a823e774af817e2748ad5ed09) the parameter to be a `string`.

This newly leads to an error being raised now, which breaks regular navigation and potentially form submissions.

Reply:
So this should be `url.toString()`?

Reply:
Feel free to open a PR to this branch!

Reply:
Opened https://github.com/ampproject/amp-wp/pull/4261